### PR TITLE
Working Login

### PR DESCRIPTION
**Pull request by Allcharles**
# Working login with baw-client

Updated login method to be compatible with baw-client

## Changes

- Re-implemented login system

## Problems

- Solution is fairly hacky because of https://github.com/QutEcoacoustics/baw-server/issues/509

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

### CODE REVIEW

File: src/app/services/baw-api/api.interceptor.service.ts [hunk 1]
```diff
@@ -46,7 +46,7 @@ export class BawApiInterceptor implements HttpInterceptor {
       return next.handle(request);
     }
 
-    // Don't add these headers to requests to cms service
+    // Don't add these headers to requests to non json requests
     if (request.responseType !== "text") {
```
Comment (atruskie) [inline src/app/services/baw-api/api.interceptor.service.ts]: the condition does not match the comment. `aplication/octet-stream` is not text... should it have these headers?

Comment (atruskie) [reply src/app/services/baw-api/api.interceptor.service.ts]: I see this is part of the responseType type's definition:

```
  responseType: 'arraybuffer' | 'blob' | 'json' | 'text'
```

Still could clear up conditon/comment

File: src/app/services/baw-api/baw-api.service.ts [hunk 2]
```diff
@@ -309,6 +309,15 @@ export abstract class BawApiService<Model extends AbstractModel> {
     return this.customFilterByForeignKey(filters, key, models, comparison);
   }
 
+  /**
+   * Concatenates path with apiRoot to form a full URL.
+   *
```
Comment (atruskie) [inline src/app/services/baw-api/baw-api.service.ts:326]: add comment on whether `path` is expected to have prefixed `/` or not

File: src/app/services/baw-api/security/security.service.ts [hunk 3]
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
...
+    function setFormData(page: any, body: URLSearchParams) {
+      const err: ApiErrorDetails = {
+        status: apiReturnCodes.unknown,
+        message: "Unable to retrieve authenticity token for login request",
+      };
+
+      // Catch wrong type of response
+      if (typeof page !== "string") {
+        throwError(err);
+      }
```
Comment (atruskie) [inline src/app/services/baw-api/security/security.service.ts]: Is this defensive or does this actually happen?

File: src/app/services/baw-api/security/security.service.ts [hunk 4]
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
...
+
+      // Extract auth token if exists
+      const token = page.match(/name="authenticity_token" value="(.+?)"/);
+      if (typeof token[1] !== "string") {
```
Comment (atruskie) [inline src/app/services/baw-api/security/security.service.ts]: a truthy check is probably sufficent here? or a check for null/undefined? regex match will always return a string

File: src/app/services/baw-api/security/security.service.ts [hunk 5]
```diff
@@ -1,19 +1,24 @@
...
+import { catchError, map, mergeMap, tap } from "rxjs/operators";
 import { ApiErrorDetails } from "../api.interceptor.service";
-import { BawApiService } from "../baw-api.service";
+import { apiReturnCodes, BawApiService } from "../baw-api.service";
 import { UserService } from "../user/user.service";
 
+function timestamp(x: number) {
+  return x;
+}
+
```
Comment (atruskie) [inline src/app/services/baw-api/security/security.service.ts]: this seems wholly redundant

File: src/app/services/baw-api/security/security.service.ts [hunk 6]
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
...
+    const formData = new URLSearchParams();
+
+    // Request login form page
+    return this.http
+      .get(this.getPath(signInEndpoint()), { responseType: "text" })
+      .pipe(
+        // Extract form data from login form
+        tap((page: string) => setFormData(page, formData)),
+        // Request api cookie
```
Comment (atruskie) [inline src/app/services/baw-api/security/security.service.ts]:
```suggestion
        // Mimic a traditional form-based sign in the get a well-formed auth cookie
        // Needed bause of https://github.com/QutEcoacoustics/baw-server/issues/509
```

File: src/app/services/baw-api/security/security.service.ts [hunk 7]
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
...
+        mergeMap(() => this.requestApiCookie(formData)),
+        // Get session user
```
Comment (atruskie) [inline src/app/services/baw-api/security/security.service.ts]:
```suggestion
        // Trade the cookie for an API auth token (mimicking old baw-client)
```

File: src/app/services/baw-api/security/security.service.ts [hunk 8]
```diff
@@ -74,6 +130,21 @@ export class SecurityService extends BawApiService<SessionUser> {
     );
   }
 
+  /**
+   * Request api auth cookie
+   */
+  private requestApiCookie(formData: URLSearchParams) {
```
Comment (atruskie) [inline src/app/services/baw-api/security/security.service.ts]:
```suggestion
  private signInWithFormData(formData: URLSearchParams) {
```

File: src/app/services/baw-api/security/security.service.ts [hunk 9]
```diff
@@ -74,6 +130,21 @@ export class SecurityService extends BawApiService<SessionUser> {
     );
   }
 
+  /**
+   * Request api auth cookie
```
Comment (atruskie) [inline src/app/services/baw-api/security/security.service.ts]:
```suggestion
   * Use the form-based sign in to authenticate. The server will issue a traditional session cookie which we will later trade for an auth token.
```

File: src/app/test/fakes/ApiErrorDetails.ts [hunk 10]
```diff
@@ -40,3 +40,22 @@ export function generateApiErrorDetails(
     info: custom?.info ?? undefined,
   };
 }
+
+export function generateApiErrorResponse(
```
Comment (atruskie) [inline src/app/test/fakes/ApiErrorDetails.ts:58]:
```suggestion
/*
 * Used to return a well constructed error object for API responses that do not conform to the standard
 */
export function generateApiErrorResponse(
```

**Review by atruskie — APPROVED** [review]
Great work. Only minor suggestions for comments recommended

File: src/app/services/baw-api/security/security.service.ts (hunk 3 above)
Comment (Allcharles) [reply src/app/services/baw-api/security/security.service.ts]: Defensive

**Pull request merged**